src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs

```diff
@@ -399,11 +481,25 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
             targetInfo,
             _compilationOptions ?? CompilationOptions.Default,
             compilationLibraries,
-            runtimeLibraries,
+            runtimeLibraries.Select(library => new RuntimeLibrary(
+                library.Library.Type,
+                library.Library.Name,
+                library.Library.Version,
+                library.Library.Hash,
+                library.Library.RuntimeAssemblyGroups,
+                library.Library.NativeLibraryGroups,
+                library.Library.ResourceAssemblies,
+                library.Library.Dependencies.Where(
+                    dependency => runtimeLibraries.Any(lib => lib.Library.Name.Equals(dependency.Name)) ||
+                    compilationLibraries.Any(lib => lib.Name.Equals(dependency.Name))).ToList(),
```

Comment:
This looks like it ends up being O(N^2) or similar.  Since I think it may not be very rare to have projects with hundreds or thousands of dependencies, we should probably use a lookup (hash or dictionary) to check if a dependency should be included rather than doing a list scan for each one.